aw-transform/src/union.rs

```diff
@@ -0,0 +1,178 @@
+use std::vec::Vec;
+
+use serde_json::{Map, Value};
+
+use aw_models::Event;
+
+fn merge_value(a: &mut Value, b: &Value) {
+    match (a, b) {
+        (&mut Value::Object(ref mut a), &Value::Object(ref b)) => {
+            for (kb, vb) in b {
+                merge_value(a.entry(kb.clone()).or_insert(Value::Null), vb);
+            }
+        }
+        (a, b) => {
+            *a = b.clone();
+        }
+    }
+}
+
+fn merge_map(map1: &mut Map<String, Value>, map2: &Map<String, Value>) {
+    for (k1, mut v1) in map1.iter_mut() {
+        if let Some(v2) = map2.get(k1) {
+            merge_value(&mut v1, &v2);
+            println!("{:?}", v1);
+        }
+    }
+    for (k2, v2) in map2.iter() {
+        if !map1.contains_key(k2) {
+            map1.insert(k2.to_string(), v2.clone());
+        }
+    }
+}
+
+/// events1 is the "master" list of events and if an event in events2
+/// intersects it the intersecting part will be removed from the original
+/// event and split into a new event and merges the data. It also differs from
+/// a normal intersection in that the part that does not intersect from the
+/// "master" events will still be kept, but if it also intersects that interval
+/// will be removed.
+///
+/// NOTE: It is technically only a union of the event1, not event2.
+///       Maybe we should improve that in the future?
+///
+/// Example:
+/// ```ignore
+///     |---------|--------------------|
+///     | events1 |[a     ][b     ]    |
+///     | events2 |    [c     ]    [d ]|
+///     | result  |[a ][ac][bc][b ]    |
+///     |---------|--------------------|
+/// ```
+pub fn union_events_split(events1: Vec<Event>, events2: &Vec<Event>) -> Vec<Event> {
+    let mut events: Vec<Event> = Vec::new();
+
+    'event1: for mut event1 in events1 {
+        let event1_endtime = event1.calculate_endtime();
+        'event2: for event2 in events2 {
```

Comment:
Instead of a nested loop one might want to step through each list under certain conditions (similarly to how we do it in aw-server-python for some transforms: `e1_i++` and `e2_i++`). 

The borrow checker would probably hate that though, and I guess the timestamp checks are pretty fast despite the worst-case `O(N^2)`.

Reply:
I don't think the borrow checker would have any issues with that, you could just use iterators which should work I think.